### DISCUSSION TITLE
Wifi interrupting PIO code

### DISCUSSION THREAD

**Discussion by ThinkTransit** [micropython/micropython#14398] [RP2040 / Pico] [answered]
I'm working on a project that sends longish (120ms) IR data bursts. The data is sent using [micropython_ir](https://github.com/peterhinch/micropython_ir) which under the hood uses PIO SM 0 to send the pulses.

Everything works fine directly from the repl however whenever the code is run in an async loop, with wifi connected and mqtt traffic flowing, the IR burst transmission gets interrupted. Sometimes there is up to 30ms pause in the IR transmission, other times it is more subtle but enough to corrupt the data.

Any tips on how to narrow down the problem? My understanding is that wifi uses the second PIO SM (4-7) so in theory PIO 0 should be able to run interrupted?

Green trace is what is sent from Pico
Yellow trace is actual pulse
![image](https://github.com/micropython/micropython/assets/74940387/6ec41cde-c46e-47e8-b9f4-9e4a4484f1b6)

**Comment by mendenm** [comment, marked as answer]
It's probably not the PIO itself that is the problem.  Looking at the code, the data is fed from an IRQ callback.  If the processor is servicing another high priority interrupt, the callback may  well be delayed.

Now that the DMA on the RP2040 is supported in micropython, this would be a great place to use it to stuff data into the PIO.  That would allow large buffers to be sent outside of program interaction, making it very  robust against interference.

**Reply by ThinkTransit** [reply]
Thanks that does make more sense, I had wrongly assumed that the full buffer was pushed into the PIO but I now see the callback is feeding the data.

DMA sounds like the way to go if the PIO can read the buffer directly from memory.

**Reply by mendenm** [reply]
Yes, the PIO works great  with DMA.  It's surprisingly straightforward.

**Reply by ThinkTransit** [reply]
Thanks any thoughts on what I'm doing wrong here?

The burst has 228 integers, converted to an array of unsigned short ints.

The PIO code loads the first item, turns on the pin, waits, loads the second item, turns off the pin, repeat.

Based on the print statements it seems like it loads a few items, usually 4 or 5 then stalls.

I'm not sure if the asm pull() command is instructing the DMA to send more data to the buffer, of if I manually need to clear the PIO buffer.

```bash
Transfers left: 223
Transfers left: 223
Transfers left: 223
```

```python
from machine import Pin, PWM
import rp2
import time
import json
from array import array


IR_TX_PIN = 17
ir_tx_pin = Pin(IR_TX_PIN)
freq= 38_000

pwm = PWM(ir_tx_pin)
pwm.freq(freq)
pwm.duty_u16(0)


burst = [3411, 1663, 464, 1236, 467, 1234, 476, 390, 540, 329, 465, 397, 470, 1235, 466, 398, 470, 397, 471, 1233, 469, 1236, 478, 389, 466, 1237, 476, 391, 475, 390, 467, 1238, 474, 1232, 471, 394, 473, 1231, 471, 1234, 470, 395, 472, 394, 473, 1233, 468, 396, 471, 396, 472, 1232, 471, 396, 469, 396, 473, 392, 475, 390, 476, 391, 465, 400, 467, 399, 468, 397, 471, 396, 471, 393, 474, 392, 474, 391, 477, 389, 467, 398, 470, 397, 469, 397, 470, 394, 472, 394, 475, 391, 475, 390, 468, 1237, 473, 393, 476, 390, 466, 399, 469, 396, 471, 401, 466, 1234, 468, 398, 469, 396, 470, 394, 474, 393, 475, 1229, 473, 1236, 467, 1236, 466, 398, 468, 396, 471, 395, 472, 394, 473, 393, 474, 1232, 471, 394, 473, 1233, 470, 394, 474, 392, 473, 392, 475, 391, 476, 389, 469, 396, 469, 398, 469, 396, 472, 393, 474, 392, 475, 391, 475, 390, 467, 399, 469, 397, 469, 397, 471, 394, 473, 393, 474, 391, 476, 390, 465, 400, 469, 398, 468, 398, 469, 395, 472, 394, 473, 392, 475, 391, 476, 388, 468, 399, 470, 395, 471, 395, 472, 394, 472, 1233, 470, 1235, 467, 420, 447, 397, 471, 394, 473, 393, 474, 1232, 469, 395, 473, 1233, 469, 395, 472, 1235, 468, 396, 472, 1234, 468, 396, 471, 0]
arr = array('H', burst)

@rp2.asm_pio(set_init=rp2.PIO.OUT_LOW)
def pulsetrain():
    wrap_target()
    pull(block)       # Manually pull data into OSR when needed
    mov(x, osr)       # Move data to 'x' for high state duration
    set(pins, 1)      # Set pin high
    label('high')
    jmp(x_dec, 'high')  # Continue high for 'x' duration
    pull(block)       # Manually pull next data for low state duration
    mov(y, osr)       # Move data to 'y'
    set(pins, 0)      # Set pin low
    label('low')
    jmp(y_dec, 'low')  # Continue low for 'y' duration
    wrap()

   
sm = rp2.StateMachine(0, pulsetrain, freq=1_000_000, set_base=ir_tx_pin)

sm.active(0)

d = rp2.DMA()
c = d.pack_ctrl(inc_write=False, size=1, treq_sel=0)

d.config(
    read=arr,
    write=sm,
    count=len(arr),
    ctrl=c
)

print("Transfers left:", d.count)
sm.active(1)
while True:
	time.sleep(1)
	print("Transfers left:", d.count)
```

**Comment by mendenm** [comment]
I think you have the wrong write address for the fifo.  Here is sample code I use:
```
        dma = self.dma = DMA()
        fifo_addr = self.pio_base + 0x10 + 4 * self.sm_idx
        dreq = 8*self.pio_idx + 0 + self.sm_idx #2.5.3.1. System DREQ Table
        ctrl = dma.pack_ctrl(size=DMA_SIZE_32, inc_write=0, inc_read=1, treq_sel=dreq)
        dma.config(write = fifo_addr, ctrl = ctrl)
```

**Reply by ThinkTransit** [reply]
Thanks for that it does look like I'm using the wrong address, I should be using this as a starting point.
```
PIO0_BASE=0x50200000
```

Turns out that the IR library was using soft interrupts and changing to hard interrupts significantly improved these timing issues.
